### PR TITLE
VPN-3362: Add plan upgrade message for monthly users

### PR DESCRIPTION
**Pull request by MattLichtenstein**
## Description

- Add new in-app messages for non-IAP subscribers on a monthly plan that become enabled on day 14 and resurfaced on day 87 of the subscription period

- Add `setTimedCallback` to the addon api to support re-evaluating javascript conditions on a timer
  - This is useful because a user who has the app launched for long periods of time will now have this message enabled at the correct time, instead of waiting for a relaunch or change in `subscriptionData` to re-evaluate the conditions.

- Add `set_subscription_start_date` inspector command to dynamically change the `createdAt` timestamp (aka start date) of the subscription for easier testing
  - This only updates the timestamp locally, so it will be overwritten every time `taskGetSubscriptionDetails` re-polls guardian

- Increase the `periodicTaskTimer` interval to fire every 5 minutes instead of every 30 seconds on the staging environment
  - This timer triggers a number of tasks, including `taskGetSubscriptionDetails`, which mad testing cumbersome

- Add a new `Subscription` badge type for in-app messages

- Add `notify` to the messages schema which determines whether a system notification should be fired for this in-app message. Default value is `true`

- Fix a bug in `subscriptiondata.cpp` where we were caching device settings instead of subscription data

Other notes: 
- The timestamp for these messages is based off of the subscription start date, not when the message was received or enabled. The timestamp for these messages will always be 14 days and 87 days respectively after the subscription start date. 
  - This will be a little janky for users who are already past this threshold as they will now have a message from a date in the past (eg once this message is deployed, a user might have a message from 10 days ago, even though it wasn't there until now)

## Reference

[VPN-3362: Release a "Plan Upgrade" in-app message to monthly-plan non-IAP subscribers](https://mozilla-hub.atlassian.net/browse/VPN-3362)
[VPN-3044: Release support for In-App Messages to generate (or not) an OS notification](https://mozilla-hub.atlassian.net/browse/VPN-3044)

### CODE REVIEW

**Review by flodolo — COMMENTED** [review]
Questions:
* Why do we need two addons that are identical in content? We're asking translators to translate these strings twice.
* Can we have better names than "14d" and "87d", given these will show up in Pontoon? Does "d" stand for "days"?

File: scripts/ci/jsonSchemas/message.json [hunk 1]
```diff
@@ -32,6 +32,10 @@
       "type": "string",
       "description": "The type of this message"
     },
+    "notify": {
+      "type": "boolean",
+      "description": "Determines whether we trigger a system notification. Default: true"
+    },
```
Comment (brizental) [inline scripts/ci/jsonSchemas/message.json:38]: This addresses https://mozilla-hub.atlassian.net/browse/VPN-3044. Please make sure to assign yourself :100:

File: src/addons/addonapi.cpp [hunk 2]
```diff
@@ -92,6 +92,32 @@ void AddonApi::initialize() {
...
+          [callback]() { callback.call(); });
+
+  timer->start();
+}
+
+void AddonApi::clearTimers() {
+  // Clear all timers for a specific addon
+  QString addonId = m_addon->id();
+  if (m_timers.contains(addonId)) {
+    m_timers.remove(addonId);
```
Comment (brizental) [inline src/addons/addonapi.cpp:117]: Should this also stop the timer or will the removal do that already?

File: addons/message_plan_upgrade_14d/conditions.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+(function(api, condition) {  
+  //Show this message after 14 days, but not after 87 days (there is a separate message for that)
+
+  //There are 2 identical (content-wise), yet separate messages because we want to give the illusion that we are 
+  //re-showing the same message 87 days in (by hiding the 14 day one and showing the 87 day one) without having to
+  //un-read or un-dismiss a message
```
Comment (brizental) [inline addons/message_plan_upgrade_14d/conditions.js:6]: Why would un-reading / un-dismissing be bad here?

File: src/addons/addonmessage.cpp [hunk 4]
```diff
@@ -61,6 +61,9 @@ Addon* AddonMessage::create(QObject* parent, const QString& manifestFileName,
   message->m_date = messageObj["date"].toInteger();
   message->planDateRetranslation();
 
+  // if "notify" is not specified in the manifest, default to true
+  message->m_shouldNotify = messageObj["notify"].toBool(true);
```
Comment (brizental) [inline src/addons/addonmessage.cpp:65]: Given this is a new API and the previous behavior used to be not showin gnotifications, should the default here be false?

File: src/constants.h [hunk 5]
```diff
@@ -172,7 +172,7 @@ CONSTEXPR(uint32_t, releaseMonitorMsec, 21600000, 4000, 0)
 
 // in milliseconds, how often we should fetch the server list, the account and
 // so on.
-CONSTEXPR(uint32_t, schedulePeriodicTaskTimerMsec, 3600000, 30000, 0)
+CONSTEXPR(uint32_t, schedulePeriodicTaskTimerMsec, 3600000, 300000, 0)
```
Comment (brizental) [inline src/constants.h:175]: Just because arithmetic is hard, please add a comment here with this time in hours or minutes.

**Review by brizental — CHANGES_REQUESTED** [review]
Please add functional tests for the new addon enabling / disabling logic.

File: src/addons/addonmessage.cpp (hunk 4 above)
Comment (MattLichtenstein) [reply src/addons/addonmessage.cpp:65]: the previous behavior was to show notifications

File: scripts/ci/jsonSchemas/message.json (hunk 1 above)
Comment (MattLichtenstein) [reply scripts/ci/jsonSchemas/message.json:38]: two for one! thanks

File: addons/message_plan_upgrade_14d/conditions.js (hunk 3 above)
Comment (MattLichtenstein) [reply addons/message_plan_upgrade_14d/conditions.js:6]: I remember investigating this route, but I don't remember why it wasn't working. This was before we had the discussion about timers. Let me reinvestigate

File: src/constants.h (hunk 5 above)
Comment (vinocher) [reply src/constants.h:175]: Is 5 minutes too long in debug for testing other scenarios? The PR description says that the previous 30 seconds made testing difficult. Could you please clarify why?

File: src/translations/strings.yaml [hunk 6]
```diff
@@ -899,6 +899,9 @@ inAppMessaging:
   surveyBadge:
     value: Survey
     comment: A badge shown in a message signifying that this message contains a survey
+  subscriptionBadge:
+    value: Subscription
+    comment: A badge shown in a message signifying that this message contains information about the users subscription
```
Comment (vinocher) [inline src/translations/strings.yaml:904]: Nit: Typo "user's"

File: src/models/subscriptiondata.cpp [hunk 7]
```diff
@@ -50,7 +50,7 @@ bool SubscriptionData::fromSettings() {
 
   logger.debug() << "Reading the subscription data from settings";
 
-  const QByteArray& json = settingsHolder->devices();
+  const QByteArray& json = settingsHolder->subscriptionData();
```
Comment (vinocher) [inline src/models/subscriptiondata.cpp:53]: Nice catch! Do you know what the symptoms of this bug was?

File: src/inspector/inspectorhandler.cpp [hunk 8]
```diff
@@ -636,7 +638,38 @@ static QList<InspectorCommand> s_commands{
...
+          qint64 newCreatedAtTimestamp = arguments[1].toLongLong();
+
+          // get sub data json from settings
+          QByteArray subscriptionData =
+              SettingsHolder::instance()->subscriptionData();
+          QJsonDocument doc = QJsonDocument::fromJson(subscriptionData);
+          QJsonObject obj = doc.object();
+          QJsonObject subObj = obj["subscription"].toObject();
+          qlonglong createdAt =
+              subObj["created"].toVariant().toLongLong() * 1000;
```
Comment (vinocher) [inline src/inspector/inspectorhandler.cpp:656]: Why is this line needed, since createdAt is overwriten in line 659?

File: docs/messages.md [hunk 9]
```diff
@@ -22,5 +22,6 @@ property object with the following properties:
 | title_comment | An optional comment to describe the meaning of the title | String | No |
 | subtitle_comment | An optional comment to describe the meaning of the subtitle | String | No |
 | date | The date the message was received (using seconds since epoch time) | Number | No |
-| badge | A label used to tag a message (options: `warning`, `critical`, `new_update`, `whats_new`, `survey` ) | String | No |
+| badge | A label used to tag a message (options: `warning`, `critical`, `new_update`, `whats_new`, `survey`, `subscription` ) | String | No |
+| notify | Should we notify the user about this message via system notifications | Boolean | No |
```
Comment (vinocher) [inline docs/messages.md:26]: SHould this indicate that default is to notify?

File: src/addons/addonapi.h [hunk 10]
```diff
@@ -27,6 +28,9 @@ class AddonApi final : public QQmlPropertyMap {
                                  const QJSValue& callback);
   Q_INVOKABLE void log(const QString& message);
 
+  Q_INVOKABLE void setTimeout(int interval, const QJSValue& callback);
```
Comment (vinocher) [inline src/addons/addonapi.h:31]: Consider if a name like scheduleCallback or setTimedCallback are better names, and change clearTimers accordingly.

**Comment by vinocher** [comment]
The PR description says:
> clearTimers should always be called before setTimeout to avoid one addon having multiple timers running (in it's current implementation)

While this may apply to these addons, isn't it conceivable that an addon may want multiple timers?

File: addons/message_plan_upgrade_87d copy/conditions.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
+(function(api, condition) {  
+  //Show this message after 87 days
+
+  //There are 2 identical (content-wise), yet separate messages because we want to give the illusion that we are 
+  //re-showing the same message 87 days in (by hiding the 14 day one and showing the 87 day one) without having to
+  //un-read or un-dismiss a message
+
+  api.connectSignal(api.subscriptionData, 'changed', () => computeCondition());
+
+  //Slight delay to as we wait for subscriptionData to initialize
```
Comment (vinocher) [inline addons/message_plan_upgrade_87d copy/conditions.js:10]: Typo "to as"

File: addons/message_plan_upgrade_87d copy/conditions.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  api.setTimeout(5000, () => computeCondition());
```
Comment (vinocher) [inline addons/message_plan_upgrade_87d copy/conditions.js:11]: Is this considered reliable to ensure that TaskGetSubscriptionDetails has run? Or should that be explicitly checked?

File: src/models/subscriptiondata.cpp (hunk 7 above)
Comment (MattLichtenstein) [reply src/models/subscriptiondata.cpp:53]: Whenever we run `TaskGetSubscriptionDetails`, we compare the incoming json with what we have stored in settings before parsing it. Here we were always comparing new subscription data json with device settings json so they were always different and we would always have to re-parse (emitting `subscriptionDataChanged`). Don't know of any bugs this caused other than wasteful signal emission/binding reevaluations

File: src/inspector/inspectorhandler.cpp (hunk 8 above)
Comment (MattLichtenstein) [reply src/inspector/inspectorhandler.cpp:656]: it's not 🙂

**Comment by MattLichtenstein** [comment]
> While this may apply to these addons, isn't it conceivable that an addon may want multiple timers?

Sure. my only reasoning was to avoid overengineering, but if you think that is a likely scenario than we can support that

File: addons/message_plan_upgrade_87d copy/conditions.js (hunk 12 above)
Comment (MattLichtenstein) [reply addons/message_plan_upgrade_87d copy/conditions.js:11]: I would say no, and I am looking for a better solution. We aren't really waiting on `TaskGetSubscriptionDetails` (if that task runs, and we have new subscription data, the `connectSignal` above will catch it), we're waiting on the subscriptionData from settings to be loaded into memory

Comment (MattLichtenstein) [reply addons/message_plan_upgrade_87d copy/conditions.js:11]: I think adding `emit changed();` [here](https://github.com/mozilla-mobile/mozilla-vpn-client/blob/e1c86b0063dbe83dbbc5b809a03fcd3eee8d14be/src/models/subscriptiondata.cpp#L59) could resolve the issue

**Comment by vinocher** [comment]
> Sure. my only reasoning was to avoid overengineering, but if you think that is a likely scenario than we can support that

I think keeping it as is by following the YAGNI principle seems correct. I was just curious.

File: src/addons/addonapi.cpp [hunk 13]
```diff
@@ -92,6 +92,35 @@ void AddonApi::initialize() {
...
+  timer->setInterval(interval);
+  m_timers.insert(m_addon->id(), timer);
+
+  connect((m_timers.value(m_addon->id())), &QTimer::timeout, this,
+          [callback]() { callback.call(); });
+
+  timer->start();
+}
+
+void AddonApi::clearTimedCallbacks() {
```
Comment (vinocher) [inline src/addons/addonapi.cpp:113]: Shouldn't the new name changes be made in the header file too. Should clearTimedCallbacks be named clearTimedCallback because only one callback is beeing cleared? Should setTimedCallback call clearTimedCallback to clear any previous callback?

Comment (MattLichtenstein) [reply src/addons/addonapi.cpp:113]: you beat me to it... working on removing the `clearTimedCallback` function altogether and just using a single timer object instead of a map. setting one timer will overwrite the other one

File: src/models/subscriptiondata.cpp [hunk 14]
```diff
@@ -50,12 +50,13 @@ bool SubscriptionData::fromSettings() {
...
   if (json.isEmpty() || !fromJsonInternal(json)) {
     return false;
   }
 
   m_rawJson = json;
+  emit changed();
```
Comment (vinocher) [inline src/models/subscriptiondata.cpp:59]: One concern is that it hasn't really changed, and does not follow the pattern in user.cpp, serverdata.cpp etc, where the signal is emitted only on an actual change.

File: src/addons/addonapi.cpp [hunk 15]
```diff
@@ -90,6 +90,22 @@ void AddonApi::initialize() {
...
+}
+
+void AddonApi::setTimedCallback(int interval, const QJSValue& callback) {
+  if (!callback.isCallable()) {
+    logger.debug() << "No callback received";
+    return;
+  }
+
+  // Creates a timer for `interval` msec before executing `callback`
+  m_timer.setInterval(interval);
```
Comment (vinocher) [inline src/addons/addonapi.cpp:104]: Does the previous timer need to be stopped or does setInterval do that automatically?

File: addons/message_plan_upgrade_14d/manifest.json
Comment (flodolo) [inline addons/message_plan_upgrade_14d/manifest.json:1]: Thanks for unifying the 2 addons.

Can we rename the folder, e.g. `message_plan_upgrade_scheduled`, or something like that, given it's not just showing after 14 days (as the name here and folder imply)?

**Review by flodolo — APPROVED** [review]

File: src/addons/addonapi.cpp [hunk 16]
```diff
@@ -90,6 +90,22 @@ void AddonApi::initialize() {
   if (s_constructorCallback) {
     s_constructorCallback(this);
   }
+
+  m_timer.setSingleShot(true);
+}
+
+void AddonApi::setTimedCallback(int interval, const QJSValue& callback) {
```
Comment (brizental) [inline src/addons/addonapi.cpp:97]:
```suggestion
void AddonApi::setTimeout(const QJSValue& callback, int interval) {
```

This design would mimic almost exactly the JavaScript setTimeout API, which I think is nice.

File: src/addons/addonapi.cpp (hunk 15 above)
Comment (brizental) [inline src/addons/addonapi.cpp:104]: Why was this changed to an interval instead of a single shot timer?

And also why was the list dropped? This prevents us from creating more than one timer. If that is the proposed design we'd need to add a check at the top of this function to only allow setting a timer when no other timer is already set... But I'd rather just make it possible to have more than one timer.

File: src/addons/addonmessage.cpp (hunk 4 above)
Comment (brizental) [reply src/addons/addonmessage.cpp:65]: LOL, I retrieve my ~dumb~ comment.

File: addons/message_plan_upgrade_14d/conditions.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        api.setTimedCallback(fourteenDaysAfterSubscriptionStarted - now, () => computeCondition());
+        condition.disable()
+      }
+      //Between 14 and 87 days into the subscription, show message
+      else if (now >= fourteenDaysAfterSubscriptionStarted && now < eightySevenDaysAfterSubscriptionStarted) {
+        api.setTimedCallback(eightySevenDaysAfterSubscriptionStarted - now, () => computeCondition())
+        api.addon.date = fourteenDaysAfterSubscriptionStarted / 1000
+        condition.enable()
+      }
+      // After 87 days into the subscription, re-surface the message (undimiss, unread, updating timestamp)
```
Comment (brizental) [inline addons/message_plan_upgrade_14d/conditions.js:39]: Question: do we care if the user has not read or dismissed the message?

**Review by brizental — CHANGES_REQUESTED** [review]
Still missing tests. Note that we can mock large jumps in current date using https://sinonjs.org/releases/latest/fake-timers/, for example.

File: src/models/subscriptiondata.cpp (hunk 14 above)
Comment (MattLichtenstein) [reply src/models/subscriptiondata.cpp:59]: I suppose it depends on how you look at it

On one hand, all of the `Q_PROPERTY`'s that rely on this `changed()` signal being emitted ARE being changed. For example, if we were to load a UI that used this data before this subscriptiondata object was initialized, that UI would not update to reflect the correct values after subscriptiondata initialization

On the other hand, the subscription data on launch is being pulled from settings, so it would be the same data as the last session. In that respect, the data hasn't change

I don't see any particular drawbacks to emitting the signal here (are there any other places we catch this signal?), but I do also dislike the inconsistency with the other data model classed. What do you think about just creating a different signal. Maybe `Initialized()`?

File: src/addons/addonapi.cpp (hunk 15 above)
Comment (MattLichtenstein) [reply src/addons/addonapi.cpp]: It's still a singleshot timer, `setInterval(msec)` just sets the timer interval, but doesn't necessarily repeat. But you're right, we don't need this as we can just pass the interval in when we start the timer.

For this particular message, only one timer is necessary, so in an effort to not overengineer and add unused complexity, I have simplified it to one overwriteable timer at a time

Comment (brizental) [reply src/addons/addonapi.cpp]: Duh, with regards to the single shot question...

Hum. Ok, let's go with the YAGNI principle. Still, this needs two things 1. unit tests 2. an error in case a timer is created while there is another timer still ongoing.

Comment (MattLichtenstein) [reply src/addons/addonapi.cpp]: if a timer is created while one is already running, we will overwrite the original timer

File: src/addons/addonapi.cpp (hunk 16 above)
Comment (vinocher) [reply src/addons/addonapi.cpp:97]: IMO, I think this is a better name than the Javascript method. Switching the parameter order to match the setTimeout looks good.

File: src/models/subscriptiondata.cpp (hunk 14 above)
Comment (vinocher) [reply src/models/subscriptiondata.cpp:59]: initialized sounds good.

File: src/addons/addonapi.cpp (hunk 16 above)
Comment (MattLichtenstein) [reply src/addons/addonapi.cpp:97]: I also like the name `setTimedCallback` a little more. I'm not sure we need to line this function up with javascript usage.

File: src/addons/addonmessage.h [hunk 18]
```diff
@@ -38,7 +38,15 @@ class AddonMessage final : public Addon {
       QString formattedDate READ formattedDate NOTIFY retranslationCompleted)
   Q_PROPERTY(Badge badge MEMBER m_badge WRITE setBadge NOTIFY badgeChanged)
 
-  enum Badge { None, Warning, Critical, NewUpdate, WhatsNew, Survey };
+  enum Badge {
+    None,
+    Warning,
+    Critical,
+    NewUpdate,
```
Comment (brizental) [inline src/addons/addonmessage.h:45]: super nit... but NewUpdate is redundant.
```suggestion
    Update,
```

File: tests/functional/addons/09_message_upgrade_to_annual_plan/message_upgrade_to_annual_plan/conditions.js [hunk 19]
```diff
@@ -0,0 +1,48 @@
+(function(api, condition) {
```
Comment (brizental) [inline tests/functional/addons/09_message_upgrade_to_annual_plan/message_upgrade_to_annual_plan/conditions.js:1]: You don't need the duplicate test addon. You can just test the real thing.

File: tests/functional/testAddons.js [hunk 20]
```diff
@@ -305,4 +305,105 @@ describe('Addons', function() {
...
+      //        "time",
+      //        true, 
+      //        '87 days after subscription created'],
+      [() => Date.now() - 1000 * 60 * 60 * 24 * 88,
+             "yesterday",
+             true, 
+             '88 days after subscription created'],
+    ];
+
+    const getNextTestCase = testCases[Symbol.iterator]();
```
Comment (brizental) [inline tests/functional/testAddons.js:339]: Fancy 💅

File: tests/functional/testAddons.js [hunk 21]
```diff
@@ -305,4 +305,105 @@ describe('Addons', function() {
...
+      }
+    }
+
+    // We call this once before all tests to set up the first test,
+    // we can't use beforeEach because that is executed after the guardian
+    // endpoints are overriden.
+    //
+    // We need to setup for the next test before it even starts for the
+    // overrides to apply.
+    setNextSubscriptionStarted(this.ctx);
```
Comment (brizental) [inline tests/functional/testAddons.js:367]: Isn't it possible to do `before` or `beforeAll`?

File: tests/unit_tests/testaddonapi.cpp [hunk 22]
```diff
@@ -186,4 +186,35 @@ void TestAddonApi::foobar() {
...
+      message, ":/addons_test/api_settimedcallback.js");
+  QVERIFY(!!a);
+
+  QTimer timer;
+
+  timer.setSingleShot(true);
+  timer.start(1000);
+
+  QObject::connect(&timer, &QTimer::timeout,
+                   [&]() { QVERIFY(a->conditionApplied()); });
```
Comment (brizental) [inline tests/unit_tests/testaddonapi.cpp:217]: Are you sure this assertion is executed before the test ends?

**Review by brizental — COMMENTED** [review]
Test are looking _beautiful_. Thank you for taking the time to add them! Left some questions and comments, nothing major.

File: src/addons/addonmessage.h (hunk 18 above)
Comment (MattLichtenstein) [reply src/addons/addonmessage.h:45]: agree. hate it. but this is not new (linter change) and is already being used by all of the "what's new" addons

File: tests/functional/testAddons.js (hunk 21 above)
Comment (MattLichtenstein) [reply tests/functional/testAddons.js:367]: the comment above says 

> // We call this once before all tests to set up the first test,
    // we can't use beforeEach because that is executed after the guardian
    // endpoints are overriden.

Comment (brizental) [reply tests/functional/testAddons.js:367]: I know but I wonderd if `before` and `beforeAll` would be called at an appropriate time.

File: tests/unit_tests/testaddonapi.cpp (hunk 22 above)
Comment (MattLichtenstein) [reply tests/unit_tests/testaddonapi.cpp:224]: nope, will add a signal spy

File: tests/functional/testAddons.js [hunk 23]
```diff
@@ -305,4 +305,103 @@ describe('Addons', function() {
...
+        await vpn.resetAddons('prod');
+
+        //Load messages
+        const loadedMessages = await vpn.messages();
+
+        //If the message is supposed to be enabled, lets check the timestamp
+        if (shouldBeAvailable) {
+            await vpn.waitForCondition(async () => (parseInt(await vpn.getMozillaProperty('Mozilla.Shared', 'MZAddonManager', 'count'), 10) > 0));
+            
+            //Check timestamp
```
Comment (brizental) [inline tests/functional/testAddons.js:381]: nit: Maybe move this comment down to line 384.

**Review by brizental — APPROVED** [review]

**Review by vinocher — APPROVED** [review]
